File: bindings/java/src/main/java/com/strling/emitters/Pcre2Emitter.java
package com.strling.emitters;

import com.strling.core.CompileResult;
import com.strling.core.IR.*;
import com.strling.core.Nodes.Flags;
import com.strling.core.STRlingCompilationError;
import com.strling.core.STRlingWarning;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * STRling PCRE2 Emitter - IR to PCRE2 Pattern String
 *
 * This class implements the emitter that transforms STRling's Intermediate
 * Representation (IR) into PCRE2-compatible regex pattern strings. The emitter:
 *   - Converts IR operations to PCRE2 syntax
 *   - Handles proper escaping of metacharacters
 *   - Manages character classes and ranges
 *   - Emits quantifiers, groups, and lookarounds
 *   - Applies regex flags as needed
 *
 * The emitter is the final stage of the compilation pipeline, producing actual
 * regex patterns that can be used with PCRE2-compatible regex engines (which
 * includes most modern regex implementations).
 */
public class Pcre2Emitter {

    /**
     * Default upper bound on AST/IR nesting depth before the emitter aborts.
     * Mirrors {@code DEFAULT_MAX_DEPTH} in the TypeScript reference (and the
     * matching constants in the C/C++/Python bindings). Tests may override
     * this via {@link #emitWithDiagnostics(IROp, Flags, int)}.
     */
    public static final int DEFAULT_MAX_DEPTH = 250;

    /**
     * Mutable context threaded through {@link #emitNode} so the depth,
     * lookbehind, and warning-collection guards can fire without polluting
     * the public API. Kept package-private; the top-level
     * {@link #emit} / {@link #emitWithDiagnostics} entry points remain
     * stateless from the caller's perspective.
     */
    static final class EmitContext {
        int depth = 0;
        int maxDepth = DEFAULT_MAX_DEPTH;
        boolean inLookbehind = false;
        final List<STRlingWarning> warnings = new ArrayList<>();
    }

    private static EmitContext newContext(int maxDepth) {
        EmitContext ctx = new EmitContext();
        ctx.maxDepth = (maxDepth > 0) ? maxDepth : DEFAULT_MAX_DEPTH;
        return ctx;
    }

    /** True iff a quantifier has an unbounded upper bound. */
    private static boolean isUnboundedQuant(IRQuant q) {
        // IRMaxBound sentinel in the wire format is the string "Inf";
        // tolerate negative ints from legacy callers as well.
        if ("Inf".equals(q.max)) return true;
        if (q.max instanceof Integer) return ((Integer) q.max) < 0;
        return false;
    }

    /** True iff a quantifier matches a variable number of characters. */
    private static boolean isVariableLengthQuant(IRQuant q) {
        // Object.equals is null-safe and value-correct for boxed Integers
        // and the "Inf" String sentinel used by the wire format.
        Object minBoxed = Integer.valueOf(q.min);
        return !minBoxed.equals(q.max);
    }

    /**
     * Mirror of {@code _isFixedLengthBody} in the TS SSOT. Returns
     * {@code true} when {@code node} consumes a fixed (statically known)
     * number of characters and is therefore safe inside a PCRE2 lookbehind.
     */
    private static boolean isFixedLengthBody(IROp node) {
        if (node instanceof IRQuant) {
            IRQuant q = (IRQuant) node;
            return !isVariableLengthQuant(q) && isFixedLengthBody(q.child);
        }
        if (node instanceof IRSeq) {
            for (IROp p : ((IRSeq) node).parts) {
                if (!isFixedLengthBody(p)) return false;
            }
            return true;
        }
        if (node instanceof IRAlt) {
            // Conservative parity with TS: every branch must be fixed-length;
            // the per-branch length-equality check is delegated to PCRE2.
            for (IROp b : ((IRAlt) node).branches) {
                if (!isFixedLengthBody(b)) return false;
            }
            return true;
        }
        if (node instanceof IRGroup) {
            return isFixedLengthBody(((IRGroup) node).body);
        }
        if (node instanceof IRLook) {
            // Nested lookarounds are zero-width, hence safe inside a lookbehind.
            return true;
        }
        // Lit, Dot, CharClass, Anchor, Backref are single- or zero-width.
        return true;
    }

    /**
     * Mirror of {@code _hasNestedUnboundedQuant} in the TS SSOT. Detects an
     * unbounded quantifier reachable from {@code child} via single-child
     * wrappers (IRGroup, single-element IRSeq) or any branch of an IRAlt.
     * Used to flag the canonical {@code (a+)+} ReDoS shape ONLY when the
     * outer quantifier is itself unbounded.
     */
    private static boolean hasNestedUnboundedQuant(IROp child) {
        if (child instanceof IRQuant) {
            return isUnboundedQuant((IRQuant) child);
        }
        if (child instanceof IRGroup) {
            return hasNestedUnboundedQuant(((IRGroup) child).body);
        }
        if (child instanceof IRSeq) {
            IRSeq seq = (IRSeq) child;
            if (seq.parts.size() == 1) return hasNestedUnboundedQuant(seq.parts.get(0));
            return false;
        }
        if (child instanceof IRAlt) {
            for (IROp b : ((IRAlt) child).branches) {
                if (hasNestedUnboundedQuant(b)) return true;
            }
        }
        return false;
    }

    private static final String REDOS_MESSAGE =
        "The pattern contains overlapping alternations or nested unbounded "
      + "quantifiers (e.g., (a+)+). This can lead to catastrophic backtracking "
      + "and exponential CPU spikes. Consider using possessive quantifiers "
      + "(++ or *+) or atomic groups to guarantee execution safety.";

    /** Append a single REDOS_RISK warning, deduplicated per emit pass. */
    private static void pushReDoSWarning(EmitContext ctx) {
        for (STRlingWarning w : ctx.warnings) {
            if ("REDOS_RISK".equals(w.getCode())) return;
        }
        ctx.warnings.add(new STRlingWarning("REDOS_RISK", REDOS_MESSAGE));
    }

    /**
     * Escapes PCRE2 metacharacters in literal strings.
     *
     * <p>Escapes characters that have special meaning in PCRE2 regex syntax when
     * used outside character classes. This ensures literal strings are matched
     * exactly as written.</p>
     *
     * @param s The literal string to escape
     * @return The escaped string safe for use in PCRE2 patterns
     */
    public static String escapeLiteral(String s) {
        // Escape PCRE2 metacharacters: . ^ $ | ( ) ? * + { } [ ] \
        StringBuilder result = new StringBuilder();
        for (char ch : s.toCharArray()) {
            switch (ch) {
                case '.':
                case '^':
                case '$':
                case '|':
                case '(':
                case ')':
                case '?':
                case '*':
                case '+':
                case '{':
                case '}':
                case '[':
                case ']':
                case '\\':
                    result.append('\\').append(ch);
                    break;
                default:
                    result.append(ch);
            }
        }
        return result.toString();
    }

    /**
     * Escapes a character for use inside [...] per PCRE2 rules.
     *
     * @param ch The character to escape
     * @return The escaped character safe for use inside character classes
     */
    public static String escapeClassChar(String ch) {
        if (ch.length() != 1) {
            throw new IllegalArgumentException("escapeClassChar expects single character");
        }
        char c = ch.charAt(0);

        // Inside [], ], \, -, ^ and [ are special and need escaping for safety.
        // ] and \ ALWAYS need escaping.
        // -, ^ and [ should be escaped to avoid ambiguity (Java's regex engine requires [ to be escaped).
        if (c == '\\' || c == ']') {
            return "\\" + c;
        }
        if (c == '[') {
            return "\\[";
        }
        if (c == '-') {
            return "\\-";
        }
        if (c == '^') {
            return "\\^";
        }

        // Handle non-printable chars / whitespace for clarity
        switch (c) {
            case '\n':
                return "\\n";
            case '\r':
                return "\\r";
            case '\t':
                return "\\t";
            case '\f':
                return "\\f";
            case '\u000B': // \v
                return "\\v";
        }

        // Handle other non-printable characters
        if (!isPrintable(c) || c < 32) {
            return String.format("\\x%02x", (int) c);
        }

        // All other characters are literal within [] including ., *, ?, [, etc.
        return String.valueOf(c);
    }

    private static boolean isPrintable(char c) {
        Character.UnicodeBlock block = Character.UnicodeBlock.of(c);
        return (!Character.isISOControl(c)) &&
                c != Character.LINE_SEPARATOR &&
                c != Character.PARAGRAPH_SEPARATOR &&
                block != null &&
                block != Character.UnicodeBlock.SPECIALS;
    }

    /**
     * Emit a PCRE2 character class. If the class is exactly one shorthand escape
     * (like \d or \p{Lu}), prefer the shorthand (with negation flipping) instead
     * of a bracketed class.
     */
    private static String emitClass(IRCharClass cc) {
        List<IRClassItem> items = cc.items;

        // --- Single-item shorthand optimization ---------------------------------
        if (items.size() == 1 && items.get(0) instanceof IRClassEscape) {
            IRClassEscape esc = (IRClassEscape) items.get(0);
            String k = esc.type;
            String prop = esc.property;

            if (k.equals("d") || k.equals("w") || k.equals("s")) {
                // Flip to uppercase negated forms when the entire class is negated.
                if (cc.negated && k.equals("d")) return "\\D";
                if (cc.negated && k.equals("w")) return "\\W";
                if (cc.negated && k.equals("s")) return "\\S";
                return "\\" + k;
            }

            if (k.equals("D") || k.equals("W") || k.equals("S")) {
                // Already-negated shorthands; flip back if the class itself is negated.
                String base = k.toLowerCase();
                return cc.negated ? ("\\" + base) : ("\\" + k);
            }

            if ((k.equals("p") || k.equals("P")) && prop != null) {
                // For \p{..}/\P{..}, flip p<->P iff exactly-negated class.
                boolean isKUpperP = k.equals("P");
                boolean useUpperP = cc.negated ^ isKUpperP;
                String use = useUpperP ? "P" : "p";
                return "\\" + use + "{" + prop + "}";
            }
        }

        // --- General case: build a bracket class --------------------------------
        StringBuilder parts = new StringBuilder();
        for (IRClassItem it : items) {
            if (it instanceof IRClassLiteral) {
                parts.append(escapeClassChar(((IRClassLiteral) it).ch));
            } else if (it instanceof IRClassRange) {
                IRClassRange range = (IRClassRange) it;
                // Escape ends of range appropriately, use unescaped - for the range operator
                parts.append(escapeClassChar(range.fromCh))
                     .append('-')
                     .append(escapeClassChar(range.toCh));
            } else if (it instanceof IRClassEscape) {
                IRClassEscape esc = (IRClassEscape) it;
                // Shorthands like \d, \p{L} are used directly
                if (esc.type.matches("[dDwWsS]")) {
                    parts.append('\\').append(esc.type);
                } else if ((esc.type.equals("p") || esc.type.equals("P")) && esc.property != null) {
                    parts.append('\\').append(esc.type).append('{').append(esc.property).append('}');
                } else {
                    // Fallback for potentially unknown escapes (shouldn't happen with valid IR)
                    parts.append('\\').append(esc.type);
                }
            } else {
                throw new UnsupportedOperationException("Unknown class item type: " + it.getClass());
            }
        }

        // Assemble the inner part
        String inner = parts.toString();
        return "[" + (cc.negated ? "^" : "") + inner + "]";
    }

    /**
     * Emit *, +, ?, {m}, {m,}, {m,n} plus optional lazy/possessive suffix.
     */
    private static String emitQuantSuffix(Object minv, Object maxv, String mode) {
        String q;

        int min = (minv instanceof Integer) ? (Integer) minv : 0;

        if (min == 0 && "Inf".equals(maxv)) {
            q = "*";
        } else if (min == 1 && "Inf".equals(maxv)) {
            q = "+";
        } else if (min == 0 && maxv.equals(1)) {
            q = "?";
        } else if (minv.equals(maxv)) {
            q = "{" + min + "}";
        } else if ("Inf".equals(maxv)) {
            q = "{" + min + ",}";
        } else {
            q = "{" + min + "," + maxv + "}";
        }

        if ("Lazy".equals(mode)) {
            q += "?";
        } else if ("Possessive".equals(mode)) {
            q += "+";
        }

        return q;
    }

    /**
     * Return true if 'child' needs a non-capturing group when quantifying.
     * Literals of length > 1, Seq, Alt, and Look typically require grouping.
     */
    private static boolean needsGroupForQuant(IROp child) {
        if (child instanceof IRCharClass || child instanceof IRDot ||
            child instanceof IRGroup || child instanceof IRBackref ||
            child instanceof IRAnchor) {
            return false;
        }
        if (child instanceof IRLit) {
            return ((IRLit) child).value.length() > 1;
        }
        // Group Alt/Look, but only group Seq if it's > 1 part
        if (child instanceof IRAlt || child instanceof IRLook) {
            return true;
        }
        if (child instanceof IRSeq) {
            return ((IRSeq) child).parts.size() > 1;
        }
        return false;
    }

    /**
     * Generate opening for group based on type.
     */
    private static String emitGroupOpen(IRGroup g) {
        if (Boolean.TRUE.equals(g.atomic)) {
            return "(?>";
        }
        if (g.capturing) {
            if (g.name != null) {
                return "(?<" + g.name + ">";
            }
            return "(";
        }
        return "(?:";
    }

    /**
     * Emit a single IR node to PCRE2 syntax.
     */
    private static String emitNode(IROp node, String parentKind, EmitContext ctx) {
        // Depth tracking surfaces the offending depth as a Signpost-pattern
        // error rather than letting the host stack overflow. Increment on
        // entry, decrement in the finally block so every return path stays
        // balanced even under exceptions.
        ctx.depth += 1;
        try {
            if (ctx.depth > ctx.maxDepth) {
                throw new STRlingCompilationError(
                    "Maximum AST depth exceeded (limit: " + ctx.maxDepth + "). "
                  + "This pattern is too deeply nested and risks host stack "
                  + "exhaustion during emission. Refactor the pattern to "
                  + "reduce nesting, or flatten capturing groups where possible.",
                    "MAX_DEPTH",
                    "pcre2"
                );
            }

            if (node instanceof IRLit) {
                return escapeLiteral(((IRLit) node).value);
            }

            if (node instanceof IRDot) {
                return ".";
            }

            if (node instanceof IRAnchor) {
                IRAnchor anchor = (IRAnchor) node;
                String at = anchor.at;
                if ("NonWordBoundary".equals(at)) {
                    return "\\B";
                }
                Map<String, String> mapping = Map.of(
                    "Start", "^",
                    "End", "$",
                    "WordBoundary", "\\b",
                    "NotWordBoundary", "\\B",
                    "AbsoluteStart", "\\A",
                    "EndBeforeFinalNewline", "\\Z",
                    "AbsoluteEnd", "\\z"
                );
                return mapping.getOrDefault(at, "");
            }

            if (node instanceof IRBackref) {
                IRBackref backref = (IRBackref) node;
                if (backref.byName != null) {
                    return "\\k<" + backref.byName + ">";
                }
                if (backref.byIndex != null) {
                    return "\\" + backref.byIndex;
                }
                return "";
            }

            if (node instanceof IRCharClass) {
                return emitClass((IRCharClass) node);
            }

            if (node instanceof IRSeq) {
                IRSeq seq = (IRSeq) node;
                StringBuilder sb = new StringBuilder();
                for (IROp p : seq.parts) {
                    sb.append(emitNode(p, "Seq", ctx));
                }
                return sb.toString();
            }

            if (node instanceof IRAlt) {
                IRAlt alt = (IRAlt) node;
                StringBuilder sb = new StringBuilder();
                for (int i = 0; i < alt.branches.size(); i++) {
                    if (i > 0) sb.append("|");
                    sb.append(emitNode(alt.branches.get(i), "Alt", ctx));
                }
                String body = sb.toString();
                if ("Seq".equals(parentKind) || "Quant".equals(parentKind)) {
                    return "(?:" + body + ")";
                }
                return body;
            }

            if (node instanceof IRQuant) {
                IRQuant quant = (IRQuant) node;
                // ReDoS guard: only flag when the *outer* quantifier is itself
                // unbounded (e.g. `(a+)+`). A bounded outer like `(a+){0,3}`
                // cannot produce exponential backtracking on its own.
                if (isUnboundedQuant(quant) && hasNestedUnboundedQuant(quant.child)) {
                    pushReDoSWarning(ctx);
                }

                String childStr = emitNode(quant.child, "Quant", ctx);
                if (needsGroupForQuant(quant.child) && !(quant.child instanceof IRGroup)) {
                    childStr = "(?:" + childStr + ")";
                }
                return childStr + emitQuantSuffix(quant.min, quant.max, quant.mode);
            }

            if (node instanceof IRGroup) {
                IRGroup group = (IRGroup) node;
                return emitGroupOpen(group) + emitNode(group.body, "Group", ctx) + ")";
            }

            if (node instanceof IRLook) {
                IRLook look = (IRLook) node;
                // Variable-length lookbehind guard: PCRE2 mandates a fixed-width
                // lookbehind body. Detect the violation here so the user sees a
                // Signpost-pattern error rather than an opaque PCRE2 compile
                // failure leaking from the runtime.
                if ("Behind".equals(look.dir) && !isFixedLengthBody(look.body)) {
                    throw new STRlingCompilationError(
                        "PCRE2 does not support variable-length lookbehinds. "
                      + "The lookbehind body contains a quantifier that makes "
                      + "its length unpredictable. Rewrite the assertion using "
                      + "a fixed-length range (e.g. `{1,8}` instead of `+`), "
                      + "or restructure the pattern using a Lookahead, or "
                      + "extract the quantified portion outside the assertion.",
                        "VLB_NOT_SUPPORTED",
                        "pcre2"
                    );
                }

                boolean wasInLookbehind = ctx.inLookbehind;
                if ("Behind".equals(look.dir)) ctx.inLookbehind = true;
                try {
                    String op;
                    if ("Ahead".equals(look.dir) && !look.neg) {
                        op = "?=";
                    } else if ("Ahead".equals(look.dir) && look.neg) {
                        op = "?!";
                    } else if ("Behind".equals(look.dir) && !look.neg) {
                        op = "?<=";
                    } else {
                        op = "?<!";
                    }
                    return "(" + op + emitNode(look.body, "Look", ctx) + ")";
                } finally {
                    ctx.inLookbehind = wasInLookbehind;
                }
            }

            throw new UnsupportedOperationException("Emitter missing for " + node.getClass());
        } finally {
            ctx.depth -= 1;
        }
    }

    /**
     * Back-compat overload for callers that already pass without a context.
     * Constructs a fresh context internally — used by the legacy
     * {@link #emit(IROp, Flags)} entry point and by direct unit tests.
     */
    private static String emitNode(IROp node, String parentKind) {
        return emitNode(node, parentKind, newContext(0));
    }


    /**
     * Build the inline prefix form expected by tests, e.g. "(?imx)"
     */
    private static String emitPrefixFromFlags(Map<String, Boolean> flags) {
        StringBuilder letters = new StringBuilder();
        if (Boolean.TRUE.equals(flags.get("ignoreCase"))) {
            letters.append('i');
        }
        if (Boolean.TRUE.equals(flags.get("multiline"))) {
            letters.append('m');
        }
        if (Boolean.TRUE.equals(flags.get("dotAll"))) {
            letters.append('s');
        }
        if (Boolean.TRUE.equals(flags.get("unicode"))) {
            letters.append('u');
        }
        if (Boolean.TRUE.equals(flags.get("extended"))) {
            letters.append('x');
        }
        return letters.length() > 0 ? "(?" + letters + ")" : "";
    }

    /**
     * Emit a PCRE2 pattern string from IR.
     *
     * If 'flags' is provided, it can be a Flags object with toDict() method.
     *
     * <p>Throws {@link STRlingCompilationError} when an emitter safety guard
     * rejects the IR (variable-length lookbehind, AST depth exceeded).
     * Diagnostic warnings (e.g. {@code REDOS_RISK}) are silently dropped
     * from this back-compat entry point; callers that need them must use
     * {@link #emitWithDiagnostics(IROp, Flags)} or the depth-aware overload.
     */
    public static String emit(IROp irRoot, Flags flags) {
        return emitWithDiagnostics(irRoot, flags, 0).getPattern();
    }

    /**
     * Emit a PCRE2 pattern string from IR without flags.
     */
    public static String emit(IROp irRoot) {
        return emit(irRoot, null);
    }

    /**
     * Like {@link #emit(IROp, Flags)}, but also returns the list of
     * {@link STRlingWarning}s collected during emission. Used by the
     * conformance test runner and any caller that wants to surface
     * {@code REDOS_RISK} (or future) warnings to the end user.
     */
    public static CompileResult emitWithDiagnostics(IROp irRoot, Flags flags) {
        return emitWithDiagnostics(irRoot, flags, 0);
    }

    /**
     * Diagnostics-bearing entry point with an optional depth-cap override.
     * Pass {@code maxDepth <= 0} to use {@link #DEFAULT_MAX_DEPTH}. The
     * override is primarily used by the conformance fixture
     * {@code tests/conformance/inputs/emitter_edges/} to provoke the depth
     * guard with a small input.
     */
    public static CompileResult emitWithDiagnostics(IROp irRoot, Flags flags, int maxDepth) {
        String prefix = "";
        if (flags != null) {
            Map<String, Boolean> flagDict = flags.toDict();
            prefix = emitPrefixFromFlags(flagDict);
        }
        EmitContext ctx = newContext(maxDepth);
        String body = emitNode(irRoot, "", ctx);
        return new CompileResult(prefix + body, ctx.warnings);
    }
}

File: bindings/swift/Sources/STRling/Emitters/PCRE2Emitter.swift
import Foundation

/// PCRE2 Emitter
///
/// Compiles the STRling AST/IR into a PCRE2-compatible regular expression string.
public class PCRE2Emitter {

    /// Default upper bound on AST/IR nesting depth before the emitter
    /// aborts. Mirrors `DEFAULT_MAX_DEPTH` in the TypeScript SSOT and
    /// the matching constants in every other binding. Tests may
    /// override via `emit(node:maxDepth:)` / `emitWithDiagnostics`.
    public static let DEFAULT_MAX_DEPTH = 250

    private static let redosMessage =
        "The pattern contains overlapping alternations or nested unbounded "
      + "quantifiers (e.g., (a+)+). This can lead to catastrophic backtracking "
      + "and exponential CPU spikes. Consider using possessive quantifiers "
      + "(++ or *+) or atomic groups to guarantee execution safety."

    /// Mutable per-emit context threaded through `emit(node:context:)`
    /// so the depth, lookbehind, and warning-collection guards can
    /// fire without polluting the public API.
    private final class EmitContext {
        var depth: Int = 0
        let maxDepth: Int
        var inLookbehind: Bool = false
        var warnings: [STRlingWarning] = []

        init(maxDepth: Int) {
            self.maxDepth = maxDepth > 0 ? maxDepth : PCRE2Emitter.DEFAULT_MAX_DEPTH
        }
    }

    public init() {}

    /// Emit PCRE2 pattern from a Node (legacy entry point).
    ///
    /// Discards any non-fatal diagnostics. Use
    /// `emitWithDiagnostics(node:maxDepth:)` to collect warnings.
    public func emit(node: Node) throws -> String {
        return try emitWithDiagnostics(node: node, maxDepth: 0).pattern
    }

    /// Emit PCRE2 pattern from a Node and surface any non-fatal
    /// diagnostics collected during emission. Pass `maxDepth <= 0` to
    /// use `DEFAULT_MAX_DEPTH`.
    public func emitWithDiagnostics(node: Node, maxDepth: Int = 0) throws -> CompileResult {
        let ctx = EmitContext(maxDepth: maxDepth)
        let pattern = try emit(node: node, context: ctx)
        return CompileResult(pattern: pattern, warnings: ctx.warnings)
    }

    // MARK: - Guard predicates (mirror the TS SSOT)

    private func isUnboundedQuant(_ q: Quant) -> Bool {
        if case .inf = q.max { return true }
        return false
    }

    private func isVariableLengthQuant(_ q: Quant) -> Bool {
        switch q.max {
        case .inf:
            return true
        case .count(let n):
            return n != q.min
        }
    }

    /// Mirror of `_isFixedLengthBody` in the TS SSOT.
    private func isFixedLengthBody(_ node: Node) -> Bool {
        switch node {
        case .quant(let q):
            return !isVariableLengthQuant(q) && isFixedLengthBody(q.child)
        case .seq(let s):
            return s.parts.allSatisfy { isFixedLengthBody($0) }
        case .alt(let a):
            return a.branches.allSatisfy { isFixedLengthBody($0) }
        case .group(let g):
            return isFixedLengthBody(g.body)
        case .look:
            return true
        default:
            return true
        }
    }

    /// Mirror of `_hasNestedUnboundedQuant` in the TS SSOT.
    private func hasNestedUnboundedQuant(_ child: Node) -> Bool {
        switch child {
        case .quant(let q):
            return isUnboundedQuant(q)
        case .group(let g):
            return hasNestedUnboundedQuant(g.body)
        case .seq(let s):
            return s.parts.count == 1 && hasNestedUnboundedQuant(s.parts[0])
        case .alt(let a):
            return a.branches.contains(where: { hasNestedUnboundedQuant($0) })
        default:
            return false
        }
    }

    private func pushReDoSWarning(_ ctx: EmitContext) {
        if ctx.warnings.contains(where: { $0.code == "REDOS_RISK" }) { return }
        ctx.warnings.append(STRlingWarning(code: "REDOS_RISK",
                                           message: PCRE2Emitter.redosMessage))
    }

    // MARK: - Depth-tracking dispatch

    private func emit(node: Node, context ctx: EmitContext) throws -> String {
        ctx.depth += 1
        defer { ctx.depth -= 1 }
        if ctx.depth > ctx.maxDepth {
            throw STRlingCompilationError(
                message: "Maximum AST depth exceeded (limit: \(ctx.maxDepth)). "
                       + "This pattern is too deeply nested and risks host stack "
                       + "exhaustion during emission. Refactor the pattern to "
                       + "reduce nesting, or flatten capturing groups where possible.",
                code: "MAX_DEPTH")
        }
        switch node {
        case .alt(let n): return try emitAlt(n, ctx: ctx)
        case .seq(let n): return try emitSeq(n, ctx: ctx)
        case .lit(let n): return emitLit(n)
        case .dot(let n): return emitDot(n)
        case .anchor(let n): return try emitAnchor(n)
        case .charClass(let n): return try emitCharClass(n)
        case .quant(let n): return try emitQuant(n, ctx: ctx)
        case .group(let n): return try emitGroup(n, ctx: ctx)
        case .backref(let n): return try emitBackref(n)
        case .look(let n): return try emitLook(n, ctx: ctx)
        }
    }

    private func emitAlt(_ node: Alt, ctx: EmitContext) throws -> String {
        let branches = try node.branches.map { try emit(node: $0, context: ctx) }
        return branches.joined(separator: "|")
    }

    private func emitSeq(_ node: Seq, ctx: EmitContext) throws -> String {
        let parts = try node.parts.map { try emit(node: $0, context: ctx) }
        return parts.joined()
    }
    
    private func emitLit(_ node: Lit) -> String {
        // Escape special PCRE2 characters
        // Note: The list of special characters might need to be comprehensive
        let specialChars = "[\\]^$.|?*+(){}"
        return node.value.map { ch in
            if specialChars.contains(ch) {
                return "\\" + String(ch)
            }
            return String(ch)
        }.joined()
    }
    
    private func emitDot(_ node: Dot) -> String {
        return "."
    }
    
    private func emitAnchor(_ node: Anchor) throws -> String {
        switch node.at {
        case "Start", "AbsoluteStart": return "^"
        case "End", "AbsoluteEnd": return "$"
        case "WordBoundary": return "\\b"
        case "NotWordBoundary": return "\\B"
        default:
            throw EmitterError.unsupportedAnchor(node.at)
        }
    }
    
    private func emitCharClass(_ node: CharClass) throws -> String {
        // --- Single-item shorthand optimization ---------------------------------
        // If the class is exactly one shorthand escape (like \d or \p{Lu}), 
        // prefer the shorthand (with negation flipping) instead of a bracketed class.
        if node.items.count == 1, let escape = node.items[0] as? ClassEscape {
            let k = escape.type
            
            // Handle d, w, s with negation flipping
            if ["d", "w", "s"].contains(k) {
                if node.negated {
                    return "\\" + k.uppercased()
                }
                return "\\" + k
            }
            
            // Handle already-negated shorthands D, W, S
            if ["D", "W", "S"].contains(k) {
                if node.negated {
                    return "\\" + k.lowercased()
                }
                return "\\" + k
            }
            
            // Handle \p{...} and \P{...}
            if ["p", "P"].contains(k), let prop = escape.property {
                let shouldNegate = node.negated != (k == "P")
                let use = shouldNegate ? "P" : "p"
                return "\\\(use){\(prop)}"
            }
        }
        
        // --- General case: build a bracket class --------------------------------
        var parts: [String] = []
        var hasHyphen = false
        
        for item in node.items {
            if let literal = item as? ClassLiteral {
                // Check if this is a hyphen literal - handle specially
                if literal.ch == "-" {
                    hasHyphen = true
                    // Don't add to parts yet - we'll add it at the beginning
                } else {
                    // Escape special characters in character class
                    if "[]\\^".contains(literal.ch) {
                        parts.append("\\" + literal.ch)
                    } else {
                        parts.append(literal.ch)
                    }
                }
            } else if let range = item as? ClassRange {
                // For ranges, we need to escape the endpoints if necessary
                let fromEscaped = "[]\\^".contains(range.fromCh) ? "\\" + range.fromCh : range.fromCh
                let toEscaped = "[]\\^".contains(range.toCh) ? "\\" + range.toCh : range.toCh
                parts.append("\(fromEscaped)-\(toEscaped)")
            } else if let escape = item as? ClassEscape {
                if let prop = escape.property {
                    parts.append("\\\(escape.type){\(prop)}")
                } else {
                    parts.append("\\\(escape.type)")
                }
            }
        }
        
        // Build the inner content with hyphen at the start if present
        let inner = hasHyphen ? "-" + parts.joined() : parts.joined()
        return "[\(node.negated ? "^" : "")\(inner)]"
    }
    
    private func emitQuant(_ node: Quant, ctx: EmitContext) throws -> String {
        // ReDoS guard: only flag when the *outer* quantifier is itself
        // unbounded (e.g. `(a+)+`). A bounded outer like `(a+){0,3}`
        // cannot produce exponential backtracking on its own.
        if isUnboundedQuant(node) && hasNestedUnboundedQuant(node.child) {
            pushReDoSWarning(ctx)
        }
        let childStr = try emit(node: node.child, context: ctx)
        // Wrap child in non-capturing group if it's a sequence or alternation to ensure correct precedence
        let needsParens: Bool
        switch node.child {
        case .seq(let s): needsParens = s.parts.count > 1
        case .alt: needsParens = true
        case .lit(let l): needsParens = l.value.count > 1
        case .quant: needsParens = true
        default: needsParens = false
        }
        
        let target = needsParens ? "(?:\(childStr))" : childStr
        
        var quantStr = ""
        switch (node.min, node.max) {
        case (0, .inf): quantStr = "*"
        case (1, .inf): quantStr = "+"
        case (0, .count(1)): quantStr = "?"
        case (let n, .count(let m)) where n == m: quantStr = "{\(n)}"
        case (let n, .inf): quantStr = "{\(n),}"
        case (let n, .count(let m)): quantStr = "{\(n),\(m)}"
        }
        
        switch node.mode {
        case "Lazy": quantStr += "?"
        case "Possessive": quantStr += "+"
        default: break
        }
        
        return target + quantStr
    }
    
    private func emitGroup(_ node: Group, ctx: EmitContext) throws -> String {
        let bodyStr = try emit(node: node.body, context: ctx)
        
        if let atomic = node.atomic, atomic {
            return "(?>\(bodyStr))"
        }
        
        if node.capturing {
            if let name = node.name {
                return "(?<\(name)>\(bodyStr))"
            }
            return "(\(bodyStr))"
        } else {
            return "(?:\(bodyStr))"
        }
    }
    
    private func emitBackref(_ node: Backref) throws -> String {
        if let name = node.byName {
            return "\\k<\(name)>"
        } else if let index = node.byIndex {
            return "\\\(index)"
        }
        throw EmitterError.invalidBackref
    }
    
    private func emitLook(_ node: Look, ctx: EmitContext) throws -> String {
        // Variable-length lookbehind guard: PCRE2 mandates a fixed-width
        // lookbehind body. Detect the violation here so the user sees a
        // Signpost-pattern error rather than an opaque PCRE2 compile
        // failure leaking from the runtime.
        if node.dir == "Behind" && !isFixedLengthBody(node.body) {
            throw STRlingCompilationError(
                message: "PCRE2 does not support variable-length lookbehinds. "
                       + "The lookbehind body contains a quantifier that makes "
                       + "its length unpredictable. Rewrite the assertion using "
                       + "a fixed-length range (e.g. `{1,8}` instead of `+`), "
                       + "or restructure the pattern using a Lookahead, or "
                       + "extract the quantified portion outside the assertion.",
                code: "VLB_NOT_SUPPORTED")
        }
        let wasInLb = ctx.inLookbehind
        if node.dir == "Behind" { ctx.inLookbehind = true }
        defer { ctx.inLookbehind = wasInLb }
        let bodyStr = try emit(node: node.body, context: ctx)
        let prefix = node.dir == "Ahead" ? "" : "<"
        let sign = node.neg ? "!" : "="
        return "(?\(prefix)\(sign)\(bodyStr))"
    }
}

public enum EmitterError: Error {
    case unsupportedAnchor(String)
    case invalidBackref
}
